### PR TITLE
Prepare Submariner env on GCP

### LINKED ISSUE #21: cloud-prepare: add support for GCP
**What would you like to be added:**

Support for automated infra configuration of GCP-based environments.

### PR DESCRIPTION
**Pull request by skeeey**
Signed-off-by: liuwei <liuweixa@redhat.com>

### CODE REVIEW

**Comment by skeeey** [comment]
refer to https://github.com/submariner-io/cloud-prepare/issues/21

**Comment by dfarrell07** [comment]
I told GH to allow the CI to run. You can see there are a number of linting errors flagged (Go, Markdown, and license headers).

Thanks for the PR!

**Comment by skeeey** [comment]
@dfarrell07 thank you

I think I fixed the lint errors, but the CI may need to be triggered again, would you help me to do this?

File: README.md [hunk 1]
```diff
@@ -63,3 +63,23 @@ In order to prepare an AWS instance, it needs to have OpenShift pre-installed an
...
 		gwDeployer, ec2.New(awsSession), infraID, region, gwInstanceType)
 ```
+
+### GCP
+
+In order to prepare a GCP instance, it needs to have OpenShift pre-installed and running.
+
+```go
+	// import "github.com/submariner-io/cloud-prepare/pkg/gcp/client" as gcpclient
+	// import "github.com/submariner-io/cloud-prepare/pkg/gcp" as cloudpreparegcp
```
Comment (mangelajo) [inline README.md]:
```suggestion
	import "github.com/submariner-io/cloud-prepare/pkg/gcp/client" as gcpclient
	import "github.com/submariner-io/cloud-prepare/pkg/gcp" as cloudpreparegcp
```

Comment (skeeey) [reply README.md]: changed

**Comment by mangelajo** [comment]
Sorry, fixed the conflict manually I hope it works.

**Review by mangelajo — APPROVED** [review]
@mkolesnik looks good to me, any comments on your side?

File: pkg/gcp/client/client.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	_, err := g.computeClient.Firewalls.Delete(g.projectID, name).Context(context.TODO()).Do()
+	return err
+}
+
+func (g *gcpClient) UpdateFirewallRule(name string, rule *compute.Firewall) error {
+	_, err := g.computeClient.Firewalls.Update(g.projectID, name, rule).Context(context.TODO()).Do()
+	return err
+}
+
+func NewClient(authJSON []byte) (Interface, error) {
```
Comment (mkolesnik) [inline pkg/gcp/client/client.go]: I think it'll be better to receive `option.ClientOption`s to pass to the client.

This way each caller can choose how to best pass credentials (and maybe other options) for itself.

File: pkg/gcp/gcp.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+// PrepareForSubmariner prepares submariner cluster environment on GCP
+func (gc *gcpCloud) PrepareForSubmariner(input api.PrepareForSubmarinerInput, reporter api.Reporter) error {
+	// create the inbound and outbound firewall rules for submariner public ports
+	ingress, egress := gc.newFirewallRules(publicPortsRuleName, input.PublicPorts, false)
+	if err := gc.openPorts(ingress, egress); err != nil {
+		return fmt.Errorf("failed to open submariner public ports: %v", err)
+	}
+
+	reporter.Succeeded("Submariner public ports are opened with firewall rules %q and %q on GCP", ingress.Name, egress.Name)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please report started and failed as well, like we do in AWS

File: pkg/gcp/gcp.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	// create the inbound firewall rule for submariner internal ports
+	internalIngress, _ := gc.newFirewallRules(internalPortsRuleName, input.InternalPorts, true)
+	if err := gc.openPorts(internalIngress); err != nil {
+		return fmt.Errorf("failed to open submariner internal ports: %v", err)
+	}
+
+	reporter.Succeeded("Submariner internal ports are opened with firewall rule %q on GCP", internalIngress.Name)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please report started and failed as well, like we do in AWS

File: pkg/gcp/gcp.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	if err := gc.client.DeleteFirewallRule(ingressName); err != nil {
+		return err
+	}
+
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner public ports are closed on GCP")
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please report started and failed as well, like we do in AWS

File: pkg/gcp/gcp.go [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(internalIngressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner internal ports are closed on GCP")
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please report started and failed as well, like we do in AWS

File: pkg/gcp/gcp.go [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+	return nil
+}
+
+func (gc *gcpCloud) newFirewallRules(name string, ports []api.PortSpec, isInternal bool) (ingress, egress *compute.Firewall) {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: I'd put all firewall related code in it's own file, for ease of maintenance of the code

File: pkg/gcp/gcp.go [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	ingress = &compute.Firewall{
+		Name:      ingressName,
+		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
+		Direction: "INGRESS",
+		Allowed:   allowedPorts,
+	}
+
+	// if ports are internal, we make the ports are accessed in internal cluster and egress is not required
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please split to 2 functions, one for external and one for internal. It'll be cleaner and simpler to understand.

Something like:
```go
func (gc *gcpCloud) newFirewallRule(name, direction string, ports []api.PortSpec) *compute.Firewall {
	allowedPorts := []*compute.FirewallAllowed{}
	for _, port := range ports {
		allowedPorts = append(allowedPorts, &compute.FirewallAllowed{
			IPProtocol: port.Protocol,
			Ports:      []string{strconv.Itoa(int(port.Port))},
		})
	}

	return &compute.Firewall{
		Name:      name,
		Network:   fmt.Sprintf("projects/%s/global/networks/%s-network", gc.projectID, gc.infraID),
		Direction: direction,
		Allowed:   allowedPorts,
	}
}

func (gc *gcpCloud) newInternalFirewallRule(name string, ports []api.PortSpec) *compute.Firewall {
	ingressName, _ := generateRuleNames(gc.infraID, name)

	rule := gc.newFirewallRule(ingressName, "INGRESS", ports)
	rule.TargetTags = []string{
		fmt.Sprintf("%s-worker", gc.infraID),
		fmt.Sprintf("%s-master", gc.infraID),
	}
	rule.SourceTags = []string{
		fmt.Sprintf("%s-worker", gc.infraID),
		fmt.Sprintf("%s-master", gc.infraID),
	}

	return rule
}

func (gc *gcpCloud) newExternalFirewallRules(name string, ports []api.PortSpec) (ingress, egress *compute.Firewall) {
	ingressName, egressName := generateRuleNames(gc.infraID, name)

	return gc.newFirewallRule(ingressName, "INGRESS", ports), gc.newFirewallRule(egressName, "EGRESS", ports)
}
```

File: pkg/gcp/gcp.go [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+	}
+}
+
+// open expected ports by creating related firewall rule
+// - if the firewall rule is not found, we will create it
+// - if the firewall rule is found and changed, we will update it
+func (gc *gcpCloud) openPorts(rules ...*compute.Firewall) error {
+	for _, rule := range rules {
+		current, err := gc.client.GetFirewallRule(rule.Name)
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 404 {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]:
```suggestion
		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusNotFound {
```

File: pkg/gcp/gcp.go [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+			}
+
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if !ruleChanged(current, rule) {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Please document why is this needed?

IIUC, it seems GCP won't change anything in this case. It seems an unnecessary optimization to me.

File: pkg/gcp/gcp.go [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+func ruleChanged(oldRule, newRule *compute.Firewall) bool {
+	if len(oldRule.Allowed) != len(newRule.Allowed) {
+		return true
+	}
+
+	if oldRule.Allowed[0].IPProtocol != newRule.Allowed[0].IPProtocol {
+		return true
+	}
+
+	return !reflect.DeepEqual(oldRule.Allowed[0].Ports, newRule.Allowed[0].Ports)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: This only checks the first allowed port, but what if there are several?

Also this function might be unnecessary based on my previous comment for `openPorts` (L152)

File: pkg/gcp/gcp.go [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+	reporter.Succeeded("Submariner internal ports are opened with firewall rule %q on GCP", internalIngress.Name)
+
+	return nil
+}
+
+// CleanupAfterSubmariner clean up submariner cluster environment on GCP
+func (gc *gcpCloud) CleanupAfterSubmariner(reporter api.Reporter) error {
+	ingressName, egressName := generateRuleNames(gc.infraID, publicPortsRuleName)
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	if err := gc.client.DeleteFirewallRule(ingressName); err != nil {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Should probably ignore a 404 here?

File: pkg/gcp/gcp.go [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
+		return err
+	}
+
+	reporter.Succeeded("Submariner public ports are closed on GCP")
+
+	// delete the inbound and outbound firewall rules to close submariner internal ports
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(internalIngressName); err != nil {
+		return err
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Should probably ignore a 404 here?

File: pkg/gcp/gcp.go [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+
+// CleanupAfterSubmariner clean up submariner cluster environment on GCP
+func (gc *gcpCloud) CleanupAfterSubmariner(reporter api.Reporter) error {
+	ingressName, egressName := generateRuleNames(gc.infraID, publicPortsRuleName)
+	// delete the inbound and outbound firewall rules to close submariner public ports
+	if err := gc.client.DeleteFirewallRule(ingressName); err != nil {
+		return err
+	}
+
+	if err := gc.client.DeleteFirewallRule(egressName); err != nil {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: Should probably ignore a 404 here?

**Review by mkolesnik — DISMISSED** [review]
Thanks, a few comments.

Also in GCP don't we need to do any special handling for the gateway nodes? it seems to be a constraint of the security model of OCP not of AWS, so I'd expect to have some similar handling necessary.

File: pkg/gcp/client/client.go (hunk 2 above)
Comment (skeeey) [reply pkg/gcp/client/client.go]: good suggestion, updated

File: pkg/gcp/gcp.go (hunk 3 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: reported

File: pkg/gcp/gcp.go (hunk 4 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: reported

File: pkg/gcp/gcp.go (hunk 5 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: reported

File: pkg/gcp/gcp.go (hunk 6 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: reported

File: pkg/gcp/gcp.go (hunk 7 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: separated

File: pkg/gcp/gcp.go (hunk 8 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: separated

File: pkg/gcp/gcp.go (hunk 9 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: using `http.StatusNotFound`

File: pkg/gcp/gcp.go (hunk 14 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: yeah, I should

File: pkg/gcp/gcp.go (hunk 13 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: yeah, I should

File: pkg/gcp/gcp.go (hunk 12 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: yeah, I should

File: pkg/gcp/gcp.go (hunk 10 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: I will try this on GCP, it GCP won't change anything, I will remove this check

Comment (skeeey) [reply pkg/gcp/gcp.go]: test on GCP, remove this check

**Comment by mangelajo** [comment]
@mkolesnik this is ready fo re-review

**Comment by mangelajo** [comment]
@skeeey I'm adding this also, I guess later on when merged we can also add to GCP, or I can update my patch if yours here merges first: https://github.com/submariner-io/cloud-prepare/pull/53

**Comment by mangelajo** [comment]
Tried to fix the conflicts on the web editor, let's see if it passes all.

**Comment by mangelajo** [comment]
@nyechiel @skitt @mkolesnik I took the liberty to target this into 0.10 / M2 and add it to the project board, otherwise it wasn't getting reviews.

File: pkg/gcp/gcp.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+	if err := gc.openPorts(ingress, egress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened public ports %q with firewall rules %q and %q on GCP",
+		formatPorts(input.PublicPorts), ingress.Name, egress.Name)
+
+	// create the inbound firewall rule for submariner internal ports
+	reporter.Started("Opening internal ports %q for intra-cluster communications on GCP", formatPorts(input.PublicPorts))
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]:
```suggestion
	reporter.Started("Opening internal ports %q for intra-cluster communications on GCP", formatPorts(input.InternalPorts))
```

File: pkg/gcp/gcp.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+	internalIngress := newInternalFirewallRule(gc.projectID, gc.infraID, input.InternalPorts)
+	if err := gc.openPorts(internalIngress); err != nil {
+		reporter.Failed(err)
+		return err
+	}
+
+	reporter.Succeeded("Opened internal ports %q with firewall rule %q on GCP",
+		formatPorts(input.PublicPorts), internalIngress.Name)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]:
```suggestion
		formatPorts(input.InternalPorts), internalIngress.Name)
```

File: pkg/gcp/gcp.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+
+// open expected ports by creating related firewall rule
+// - if the firewall rule is not found, we will create it
+// - if the firewall rule is found and changed, we will update it
+func (gc *gcpCloud) openPorts(rules ...*compute.Firewall) error {
+	for _, rule := range rules {
+		_, err := gc.client.GetFirewallRule(gc.projectID, rule.Name)
+		if isGCPNotFoundError(err) {
+			if err := gc.client.InsertFirewallRule(gc.projectID, rule); err != nil {
+				fmt.Println(rule)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: This print out intentional?

File: pkg/gcp/gcp.go [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if err := gc.client.UpdateFirewallRule(gc.projectID, rule.Name, rule); err != nil {
+			fmt.Println(rule)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: This print out intentional?

File: pkg/gcp/gcp.go [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+	return nil
+}
+
+func formatPorts(ports []api.PortSpec) string {
+	portStrs := []string{}
+	for _, port := range ports {
+		portStrs = append(portStrs, fmt.Sprintf("%d/%s", port.Port, port.Protocol))
+	}
+
+	return strings.Join(portStrs, ",")
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]:
```suggestion
	return strings.Join(portStrs, ", ")
```

File: pkg/gcp/gcp.go [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+func isGCPNotFoundError(err error) bool {
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: It would make more sense to put this in client.go

File: pkg/gcp/gcp.go [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+	reporter.Started("Deleting ingress firewall rule %q on GCP", internalPortsRuleName)
+	internalIngressName, _ := generateRuleNames(gc.infraID, internalPortsRuleName)
+	if err := gc.client.DeleteFirewallRule(gc.projectID, internalIngressName); err != nil {
+		if !isGCPNotFoundError(err) {
+			reporter.Failed(err)
+			return err
+		}
+	}
+
+	reporter.Succeeded("Deleted ingress firewall rule %q on GCP", internalPortsRuleName)
```
Comment (mkolesnik) [inline pkg/gcp/gcp.go]: This code is pretty repetitive with the other 2 calls which also delete public rules, it seems to me you can extract this to a function, eg

```go
func (gc *gcpCloud) deleteFirewallRule(name string) error {
	reporter.Started("Deleting firewall rule %q on GCP", name)
	if err := gc.client.DeleteFirewallRule(gc.projectID, name); err != nil {
		if !isGCPNotFoundError(err) {
			reporter.Failed(err)
			return err
		}
	}

	reporter.Succeeded("Deleted firewall rule %q on GCP", name)
}
```

You can then call it for each of the public ingress/egress and the private ingress

**Review by mkolesnik — CHANGES_REQUESTED** [review]
Thanks for working on this, a few more comments

File: pkg/gcp/gcp.go (hunk 15 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: my carelessness... updated

File: pkg/gcp/gcp.go (hunk 16 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: my carelessness... updated

File: pkg/gcp/gcp.go (hunk 17 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: yes, it's for debug, my carelessness, removed

File: pkg/gcp/gcp.go (hunk 18 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: yes, it's for debug, my carelessness, removed

File: pkg/gcp/gcp.go (hunk 19 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: updated

File: pkg/gcp/gcp.go (hunk 20 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: move to client.go

File: pkg/gcp/gcp.go (hunk 21 above)
Comment (skeeey) [reply pkg/gcp/gcp.go]: good suggestion, updated

File: README.md [hunk 22]
```diff
@@ -63,3 +63,33 @@ In order to prepare an AWS instance, it needs to have OpenShift pre-installed an
 	cloud := cloudprepareaws.NewCloud(
 		gwDeployer, ec2.New(awsSession), infraID, region, gwInstanceType)
 ```
+
+### GCP
+
+In order to prepare a GCP instance, it needs to have OpenShift pre-installed and running.
```
Comment (mangelajo) [inline README.md:69]: I hope in the future we can find community support to extend this for GKE, or other distros also.

**Review by mangelajo — APPROVED** [review]

**Comment by skeeey** [comment]
> Thanks, a few comments.
> 
> Also in GCP don't we need to do any special handling for the gateway nodes? it seems to be a constraint of the security model of OCP not of AWS, so I'd expect to have some similar handling necessary.

For GCP, we don’t do any special handling for nodes, we just need to open the ports on the firewall, then any worker node can be used as gateway node

**Comment by nyechiel** [comment]
@skeeey this has a conflict now, can you please review again? We are wrapping-up the 0.10 release, and I think we need to try and merge this before. Thanks!

**Comment by skeeey** [comment]
@mkolesnik would you help me merge this pr firstly? then I will base on this to create another pr to handle the dedicated gateway part as we discussed

**Comment by dfarrell07** [comment]
@mkolesnik Can you re-review please?

**Review by mkolesnik — APPROVED** [review]
Looks good, let's merge and I'll adjust #72 accordingly to also refactor this code

**Review by mkolesnik — CHANGES_REQUESTED** [review]
I see you added some unit tests but they're not `Ginkgo` based.

In the various projects we only write Ginkgo tests, so I suggest you either remove them or convert them to Ginkgo before we merge this patch.

Attempting to run the unit tests with `make unit` fails entirely so they're not useful as of now anyhow.

**Comment by skeeey** [comment]
@mkolesnik test is removed

**Review by mangelajo — APPROVED** [review]

**Review by mkolesnik — APPROVED** [review]
